clippy_lints/src/strings_early.rs

```diff
@@ -0,0 +1,68 @@
+use crate::rustc::lint::{EarlyContext, EarlyLintPass, LintArray, LintPass};
+use crate::rustc::{declare_tool_lint, lint_array};
+use crate::utils::span_lint_and_sugg;
+
+/// **What it does:** Checks for the `as_bytes` method called on string literals
+/// that contain only ASCII characters.
+///
+/// **Why is this bad?** Byte string literals (e.g. `b"foo"`) can be used
+/// instead. They are shorter but less discoverable than `as_bytes()`.
+///
+/// **Known Problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// let bs = "a byte string".as_bytes();
+/// ```
+declare_clippy_lint! {
+    pub STRING_LIT_AS_BYTES,
+    style,
+    "calling `as_bytes` on a string literal instead of using a byte string literal"
+}
+
+#[derive(Copy, Clone)]
+pub struct StringLitAsBytes;
+
+impl LintPass for StringLitAsBytes {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(STRING_LIT_AS_BYTES)
+    }
+}
+
+impl EarlyLintPass for StringLitAsBytes {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, e: &crate::syntax::ast::Expr) {
+        use crate::syntax::ast::{ExprKind, LitKind};
+        use crate::utils::{in_macro, snippet};
+
+        if let ExprKind::MethodCall(ref path, ref args) = e.node {
+            if path.ident.name == "as_bytes" {
+                if let ExprKind::Lit(ref lit) = args[0].node {
+                    if let LitKind::Str(ref lit_content, _) = lit.node {
+                        if lit_content.as_str().chars().all(|c| c.is_ascii()) && !in_macro(args[0].span) {
+                            span_lint_and_sugg(
+                                cx,
+                                STRING_LIT_AS_BYTES,
+                                e.span,
+                                "calling `as_bytes()` on a string literal",
+                                "consider using a byte string literal instead",
+                                format!("b{}", snippet(cx, args[0].span, r#""foo""#)),
+                            );
+                        }
+                    }
+                } else if let ExprKind::Mac(ref mac) = args[0].node {
+                    if mac.node.path == "include_str" {
+                        cx.sess.err(&format!("{:?}", mac));
+                        span_lint_and_sugg(
+                            cx,
+                            STRING_LIT_AS_BYTES,
+                            e.span,
+                            "calling `as_bytes()` on a include_str macro",
+                            "consider using include_bytes instead",
+                            format!("include_str{}", snippet(cx, mac.span, "include_bytes")),
```

Comment:
This part is broken, I'm having trouble figuring out how to decompose the macro in order to swap out the include_str with include_bytes. Also I'm assuming that we shouldn't be recommending `bstringify!` instead of `stringify!` because I couldn't find any reference to such a macro in the documentation, so the intent here is to only lint for include_str -> include_bytes when its `macro!().as_bytes()`

Reply:
Yeah this should just trigger on the `include_str!` macro.

I think the arguments of the macro are in `mac.node.tts`. You should be able to format this with
```rust
format!("include_bytes!({})", mac.node.stream());
```
`mac.node.stream()` gives you a `TokenStream`, which implements the `Display` trait. [Documentation](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/struct.Mac_.html)